File: algorithms/1854.maximum-population-year/solution.go
func maximumPopulation(logs [][]int) int {
	m := map[int]int{}

	maxPopYear := logs[0][0]
	maxPop := 0
	for _, log := range logs {
		for year := log[0]; year < log[1]; year++ {
			if _, ok := m[year]; ok {
				m[year] += 1
			} else {
				m[year] = 1
			}

			if maxPop < m[year] {
				maxPop = m[year]
				maxPopYear = year
			} else if maxPop == m[year] {
				if year < maxPopYear {
					maxPopYear = year
				}
			}
		}
	}

	return maxPopYear
}

File: algorithms/1854.maximum-population-year/solution.rs
use std::collections::HashMap;

impl Solution {
    pub fn maximum_population(logs: Vec<Vec<i32>>) -> i32 {
        let mut map: HashMap<i32, i32> = HashMap::new();

        let mut max_pop = 0;
        let mut max_pop_year = logs[0][0];
        for log in logs {
            for year in log[0]..log[1] {
                *map.entry(year).or_insert(0) += 1;

                if max_pop < map[&year] {
                    max_pop = map[&year];
                    max_pop_year = year;
                } else if max_pop == map[&year] {
                    if year < max_pop_year {
                        max_pop_year = year;
                    }
                }
            }
        }

        max_pop_year
    }
}
